cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -565,106 +663,107 @@ func (scaleSet *ScaleSet) Nodes() ([]cloudprovider.Instance, error) {
 		return scaleSet.instanceCache, nil
 	}
 
-	klog.V(4).Infof("Nodes: starts to get VMSS VMs")
-	splay := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(scaleSet.instancesRefreshJitter + 1)
-	lastRefresh := time.Now().Add(-time.Second * time.Duration(splay))
-
-	orchestrationMode, err := scaleSet.getOrchestrationMode()
+	// Forcefully updating the instanceCache as the instanceCacheSize didn't match curSize or cache is invalid.
+	err = scaleSet.updateInstanceCache()
 	if err != nil {
-		klog.Errorf("failed to get information for VMSS: %s, error: %v", scaleSet.Name, err)
 		return nil, err
 	}
 
-	klog.V(4).Infof("VMSS: orchestration Mode %s", orchestrationMode)
-
-	if orchestrationMode == compute.Uniform {
-		err := scaleSet.buildScaleSetCache(lastRefresh)
-		if err != nil {
-			return nil, err
-		}
-
-	} else if orchestrationMode == compute.Flexible {
-		if scaleSet.manager.config.EnableVmssFlex {
-			err := scaleSet.buildScaleSetCacheForFlex(lastRefresh)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			return nil, fmt.Errorf("vmss - %q with Flexible orchestration detected but 'enableVmssFlex' feature flag is turned off", scaleSet.Name)
-		}
-
-	} else {
-		return nil, fmt.Errorf("Failed to determine orchestration mode for vmss %q", scaleSet.Name)
-	}
-
 	klog.V(4).Infof("Nodes: returns")
 	return scaleSet.instanceCache, nil
 }
 
-func (scaleSet *ScaleSet) buildScaleSetCache(lastRefresh time.Time) error {
-	vms, rerr := scaleSet.GetScaleSetVms()
+// buildScaleSetCacheForFlex is used by orchestrationMode == compute.Flexible
+func (scaleSet *ScaleSet) buildScaleSetCacheForFlex() error {
+	klog.V(3).Infof("buildScaleSetCacheForFlex: resetting instance Cache for scaleSet %s",
+		scaleSet.Name)
+	splay := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(scaleSet.instancesRefreshJitter + 1)
+	lastRefresh := time.Now().Add(-time.Second * time.Duration(splay))
+
+	vms, rerr := scaleSet.GetFlexibleScaleSetVms()
 	if rerr != nil {
 		if isAzureRequestsThrottled(rerr) {
 			// Log a warning and update the instance refresh time so that it would retry after cache expiration
-			klog.Warningf("GetScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
+			klog.Warningf("GetFlexibleScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
 			scaleSet.lastInstanceRefresh = lastRefresh
 			return nil
 		}
 		return rerr.Error()
 	}
 
-	scaleSet.instanceCache = buildInstanceCache(vms)
+	scaleSet.instanceCache = buildInstanceCacheForFlex(vms)
 	scaleSet.lastInstanceRefresh = lastRefresh
 
 	return nil
 }
 
-func (scaleSet *ScaleSet) buildScaleSetCacheForFlex(lastRefresh time.Time) error {
```

Comment:
Should be compared against the method of the same name, not `buildScaleSetCacheForUniform()`

Reply:
It looks like `buildInstanceCache()` is completely replaced by `buildInstanceCacheForFlex()`. Non-flex cases probably don't need a separate method and can stay in `buildScaleSetCacheForUniform()`

Reply:
`buildInstanceCacheForFlex()` is in fact similar to `buildInstanceCache()`. See below.